### PR TITLE
Solo thierrymoudiki release3

### PR DESCRIPTION
**Pull request by thierrymoudiki**

### CODE REVIEW

File: app/controllers/suggestion_controller.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  puts @suggestion
+  if @suggestion
+    erb :"/suggestions/show"
+  else
+    "Suggestion not found"
+  end
+end
+
+post "/suggestions" do
+  @suggestion = Suggestion.new(title: params[:suggestion][:title],
```
Comment (panw) [inline app/controllers/suggestion_controller.rb]: 1. Look into utilizing namespacing in your form, so you could nest suggestion title, description, user to be like this

```
{ 
  suggestion: {
    title: '',
    description: '',
    user_id: 1
  }
}
```
You could then pass `Suggestion.new(params[:suggestion])

2. The variable you defined also doesn't have to be an instance variable.

File: app/controllers/user_controller.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+get '/users/new' do
+  erb :'/users/new'
+end
+
+get '/users/login' do
+  @user = User.find_by(id: session[:user_id])
+  if @user
+     # User publications
+    @user_own_suggestions = Suggestion.where(user_id: session[:user_id])
```
Comment (panw) [inline app/controllers/user_controller.rb:27]: Could do `user.suggestions` instead

File: app/views/suggestions/index.erb [hunk 3]
```diff
@@ -0,0 +1,128 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>Bootstrap Example</title>
```
Comment (panw) [inline app/views/suggestions/index.erb:4]: Don't work utilizing bootstrap until you have CRUD better.

File: public/js/application.js [hunk 4]
```diff
@@ -4,4 +4,38 @@ $(document).ready(function() {
...
+    $("#login-form").show();
+    });
+
+  $("a[href = '#delete']").on("click", function(event) {
+    event.preventDefault();
+    $(this).parent().find('.edit-suggestion').hide();
+    $(this).parent().find('.delete-suggestion').show();
+    });
+
+  $("a[href = '#edit']").on("click", function(event) {
```
Comment (panw) [inline public/js/application.js:30]: The focus is on pure CRUD, so I would not add in jQuery yet.

File: db/migrate/20171011100324_create_user.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class CreateUser < ActiveRecord::Migration[5.1]
+  def up
+    create_table :users do |t|
+      t.column :name, :string, unique: true
+      t.column :password_digest, :string
```
Comment (philsof) [inline db/migrate/20171011100324_create_user.rb]: 👍  reversible migration. let's get some `null: false` constraints where appropriate also, across all of your migration files.

File: db/seeds.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+User.delete_all
+Suggestion.delete_all
+
+5.times do
+  user = User.create!(name: Faker::Name.name, password: Faker::Lorem.word)
+  suggestion = Suggestion.create!(title: Faker::Lorem.word,
+    description: Faker::Lorem.sentence, user_id: user.id)
```
Comment (philsof) [inline db/seeds.rb:7]: 👍 dynamically associating users to suggestions in seed file

File: app/models/suggestion.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+class Suggestion < ApplicationRecord
+  # Remember to create a migration!
+  belongs_to :user
+  validates :title, presence: true
+  validates :description, presence: true
+  validates :user_id, presence: true
```
Comment (philsof) [inline app/models/suggestion.rb]: Even better:
```ruby
validates :user, presence: true
```
Question: why is this better?

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  # Remember to create a migration!
+  has_many :suggestions
+  validates :name, presence: true, uniqueness: true
+  validates :password, presence: true
+
+  has_secure_password
```
Comment (philsof) [inline app/models/user.rb:7]: 👍

File: app/controllers/suggestion_controller.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    redirect "/suggestions"
+  else
+    "Cannot add the suggestion"
+  end
+end
+
+post "/suggestions/login" do
+  @suggestion = Suggestion.new(title: params[:suggestion][:title],
+    description: params[:suggestion][:description], user_id: session[:user_id])
+  if @suggestion.save
```
Comment (philsof) [inline app/controllers/suggestion_controller.rb:41]: Why is the route on this action `/suggestion/login`? (Looks like this action creates a new suggestion.)

File: app/controllers/suggestion_controller.rb [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  else
+    "Cannot add the suggestion"
+  end
+end
+
+delete "/suggestions/:id" do
+  @suggestion = Suggestion.find_by(id: params[:id])
+  # protect the suggestion
+  if @suggestion.user_id.to_i == session[:user_id].to_i
+    @suggestion.destroy
```
Comment (philsof) [inline app/controllers/suggestion_controller.rb:52]: 👍 protecting the destroy

File: app/controllers/suggestion_controller.rb [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  else
+    "Suggestion not found"
+  end
+end
+
+put "/suggestions/:id" do
+  @suggestion = Suggestion.find_by(id: params[:id])
+  # protect the suggestion
+  if @suggestion.user_id.to_i == session[:user_id].to_i
+    @suggestion.assign_attributes(title: params[:suggestion][:title], description: params[:suggestion][:description], user_id: session[:user_id])
```
Comment (philsof) [inline app/controllers/suggestion_controller.rb:63]: 👍

File: app/controllers/user_controller.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
+post '/users/signup' do
+  @user = User.new(:name => params[:user][:name], :password => params[:user][:password])
```
Comment (philsof) [inline app/controllers/user_controller.rb]: FYI the rocket `=>` is starting to go out of style

File: app/controllers/user_controller.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  if @user.save
+    erb :'/users/new'
+  else
+    "Cannot sign up, username already exists"
```
Comment (philsof) [inline app/controllers/user_controller.rb]: Let's get accurate error messaging here, since the reasons for user create failure include possibilities besides username already existing (e.g. blank password)

File: app/controllers/user_controller.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    # Other users' publications
+    @user_other_suggestions = Suggestion.where.not(user_id: session[:user_id])
+    erb :'/users/login'
+  else
+    "Cannot log in, user not registered"
+  end
+end
+
+get '/users/logout' do
+  session[:user_id] = nil
```
Comment (philsof) [inline app/controllers/user_controller.rb]: Best to clear the entire session hash, for security reasons (because Sinatra creates additional key/value pairs in the session hash that should also be deleted):
```ruby
session.clear # deletes everything in the hash, but not the hash itself
```

File: app/views/index.erb [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+<head>
+  <title>Bootstrap Example</title>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
+  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
+  <style>
+    /* Remove the navbar's default margin-bottom and rounded borders */
+    .navbar {
```
Comment (philsof) [inline app/views/index.erb:12]: Move all styling to separate CSS files to maintain DRY principle and for scalability

**Review by philsof — COMMENTED** [review]
@thierrymoudiki see inline notes

File: app/controllers/user_controller.rb (hunk 2 above)
Comment (thierrymoudiki) [reply app/controllers/user_controller.rb:27]: Totally equivalent? Or better? Is there any equivalent for Suggestion.where.not?

File: app/controllers/suggestion_controller.rb (hunk 9 above)
Comment (thierrymoudiki) [reply app/controllers/suggestion_controller.rb:41]: Yes, should be in suggestions/new. 
I'll see how to modify this, and keep the same UX.

File: app/controllers/user_controller.rb (hunk 12 above)
Comment (thierrymoudiki) [reply app/controllers/user_controller.rb]: Modified in new push.

File: app/controllers/user_controller.rb (hunk 13 above)
Comment (thierrymoudiki) [reply app/controllers/user_controller.rb]: True. I added more cases now. Not sure how to specifically check that the username already exists.

File: app/controllers/user_controller.rb (hunk 14 above)
Comment (thierrymoudiki) [reply app/controllers/user_controller.rb]: Corrected now.

File: app/models/suggestion.rb (hunk 7 above)
Comment (thierrymoudiki) [reply app/models/suggestion.rb]: Corrected now.
Because it checks if the join with an existing user in table User can be made?

File: app/views/index.erb (hunk 15 above)
Comment (thierrymoudiki) [reply app/views/index.erb:12]: Styles now moved in public/css

File: db/migrate/20171011100324_create_user.rb (hunk 5 above)
Comment (thierrymoudiki) [reply db/migrate/20171011100324_create_user.rb]: Done.
